torch_xla/csrc/ops/ops.cpp

```diff
@@ -791,6 +792,47 @@ NodePtr LogicalAnd(const Value& input, const Value& other) {
       std::move(lower_fn));
 }
 
+NodePtr NanToNum(const Value& input, c10::optional<double> nan,
+                 c10::optional<double> posinf, c10::optional<double> neginf) {
+  auto lower_fn = [nan, posinf, neginf](const Node& node,
+                                        LoweringContext* loctx) -> XlaOpVector {
+    xla::XlaOp xla_input = loctx->GetOutputOp(node.operand(0));
+    const xla::Shape& input_shape = XlaHelpers::ShapeOfXlaOp(xla_input);
+    auto element_type = input_shape.element_type();
+
+    if (!xla::primitive_util::IsFloatingPointType(element_type)) {
+      return node.ReturnOp(xla_input, loctx);
+    }
+
+    xla::XlaOp nan_replacement =
+        nan.has_value()
+            ? XlaHelpers::ScalarValue(*nan, element_type, xla_input.builder())
+            : xla::Zero(xla_input.builder(), element_type);
+    xla::XlaOp pos_inf_replacement =
+        posinf.has_value()
+            ? XlaHelpers::ScalarValue(*posinf, element_type,
+                                      xla_input.builder())
+            : xla::MaxFiniteValue(xla_input.builder(), element_type);
+    xla::XlaOp neg_inf_replacement =
+        posinf.has_value()
+            ? XlaHelpers::ScalarValue(*neginf, element_type,
+                                      xla_input.builder())
+            : xla::MinFiniteValue(xla_input.builder(), element_type);
+
+    xla::XlaOp result =
+        xla::Select(xla::IsNan(xla_input), nan_replacement,
+                    xla::Select(xla::IsPosInf(xla_input), pos_inf_replacement,
+                                xla::Select(xla::IsNegInf(xla_input),
+                                            neg_inf_replacement, xla_input)));
+    return node.ReturnOp(result, loctx);
+  };
+  using hash_func = std::hash<c10::optional<double>>;
+  auto hash_value = xla::util::MHash(hash_func{}(nan), hash_func{}(posinf),
+                                     hash_func{}(neginf));
+  return GenericOp(OpKind(at::aten::nan_to_num), {input}, input.shape(),
+                   std::move(lower_fn), /*num_outputs=*/1, hash_value);
```

Comment:
`xla::util::MHash(nan, posinf, neginf)` should just work after you get rid of optional here

Reply:
It looks like `xla::util::MHash` does not work with `at::Scalar` directly.